File: Squatch Sports Basketball Training WatchOS Companion/AppDataStore.swift
import SwiftUI
import Foundation
import Combine

struct WorkoutSession: Identifiable, Codable {
    let id: UUID
    let drill: String
    let makes: Int
    let misses: Int
    let swishes: Int
    let attempts: Int
    let startDate: Date
    let endDate: Date

    init(
        id: UUID = UUID(),
        drill: String,
        makes: Int,
        misses: Int,
        swishes: Int,
        attempts: Int,
        startDate: Date,
        endDate: Date
    ) {
        self.id = id
        self.drill = drill
        self.makes = makes
        self.misses = misses
        self.swishes = swishes
        self.attempts = attempts
        self.startDate = startDate
        self.endDate = endDate
    }

    var percentage: Int {
        guard attempts > 0 else { return 0 }
        return Int((Double(makes) / Double(attempts)) * 100)
    }
}

final class AppDataStore: ObservableObject {
    private let defaults: UserDefaults
    private enum Keys {
        static let sessions = "squatch.sessions"
        static let dailyShotGoal = "squatch.dailyShotGoal"
        static let weeklySessionGoal = "squatch.weeklySessionGoal"
        static let targetFGGoal = "squatch.targetFGGoal"
    }

    @Published var sessions: [WorkoutSession] {
        didSet { persistSessions() }
    }

    @Published var dailyShotGoal: Int {
        didSet { defaults.set(dailyShotGoal, forKey: Keys.dailyShotGoal) }
    }

    @Published var weeklySessionGoal: Int {
        didSet { defaults.set(weeklySessionGoal, forKey: Keys.weeklySessionGoal) }
    }

    @Published var targetFGGoal: Int {
        didSet { defaults.set(targetFGGoal, forKey: Keys.targetFGGoal) }
    }

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults

        if let data = defaults.data(forKey: Keys.sessions),
           let decoded = try? JSONDecoder().decode([WorkoutSession].self, from: data) {
            self.sessions = decoded
        } else {
            self.sessions = []
        }

        self.dailyShotGoal = (defaults.object(forKey: Keys.dailyShotGoal) as? Int) ?? 200
        self.weeklySessionGoal = (defaults.object(forKey: Keys.weeklySessionGoal) as? Int) ?? 5
        self.targetFGGoal = (defaults.object(forKey: Keys.targetFGGoal) as? Int) ?? 75
    }

    private func persistSessions() {
        if let data = try? JSONEncoder().encode(sessions) {
            defaults.set(data, forKey: Keys.sessions)
        }
    }

    func addSession(_ session: WorkoutSession) {
        sessions.insert(session, at: 0)
    }

    var totalMakes: Int {
        sessions.reduce(0) { $0 + $1.makes }
    }

    var totalAttempts: Int {
        sessions.reduce(0) { $0 + $1.attempts }
    }

    var totalMisses: Int {
        sessions.reduce(0) { $0 + $1.misses }
    }

    var totalSwishes: Int {
        sessions.reduce(0) { $0 + $1.swishes }
    }

    var shootingPercentage: Int {
        guard totalAttempts > 0 else { return 0 }
        return Int((Double(totalMakes) / Double(totalAttempts)) * 100)
    }

    var sessionsCompleted: Int {
        sessions.count
    }

    var favoriteDrill: String {
        let grouped = Dictionary(grouping: sessions, by: { $0.drill })
        return grouped.max(by: { $0.value.count < $1.value.count })?.key ?? "No sessions yet"
    }

    var todayMakes: Int {
        sessions
            .filter { Calendar.current.isDateInToday($0.endDate) }
            .reduce(0) { $0 + $1.makes }
    }

    var todayAttempts: Int {
        sessions
            .filter { Calendar.current.isDateInToday($0.endDate) }
            .reduce(0) { $0 + $1.attempts }
    }

    var weeklySessionsCompleted: Int {
        sessions.filter { session in
            guard let weekInterval = Calendar.current.dateInterval(of: .weekOfYear, for: Date()) else {
                return false
            }
            return weekInterval.contains(session.endDate)
        }.count
    }

    var bestSession: WorkoutSession? {
        sessions.max { $0.percentage < $1.percentage }
    }

    var mostRecentSession: WorkoutSession? {
        sessions.sorted(by: { $0.endDate > $1.endDate }).first
    }

    var currentFocus: String {
        if shootingPercentage < targetFGGoal {
            return "Improve shot consistency"
        } else if weeklySessionsCompleted < weeklySessionGoal {
            return "Complete more sessions this week"
        } else {
            return "Keep building momentum"
        }
    }

    var todayGoalProgress: Double {
        guard dailyShotGoal > 0 else { return 0 }
        return min(Double(todayAttempts) / Double(dailyShotGoal), 1.0)
    }

    func clearAllSessions() {
        sessions.removeAll()
    }
}

File: android/app/src/main/java/com/squatchsports/training/data/AppDataStore.kt
package com.squatchsports.training.data

import android.content.Context
import androidx.core.content.edit
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.setValue
import com.squatchsports.training.shared.WorkoutSession
import org.json.JSONArray
import org.json.JSONObject
import java.util.Calendar

class AppDataStore(context: Context) {
    private val preferences = context.getSharedPreferences("squatch_data", Context.MODE_PRIVATE)

    val sessions = mutableStateListOf<WorkoutSession>()

    var dailyShotGoal by mutableIntStateOf(preferences.getInt(KEY_DAILY_SHOT_GOAL, 200))
        private set
    var weeklySessionGoal by mutableIntStateOf(preferences.getInt(KEY_WEEKLY_SESSION_GOAL, 5))
        private set
    var targetFgGoal by mutableIntStateOf(preferences.getInt(KEY_TARGET_FG_GOAL, 75))
        private set

    init {
        sessions.addAll(readSessions())
    }

    fun addSession(session: WorkoutSession) {
        sessions.add(0, session)
        persistSessions()
    }

    fun clearAllSessions() {
        sessions.clear()
        persistSessions()
    }

    val totalMakes: Int get() = sessions.sumOf { it.makes }
    val totalAttempts: Int get() = sessions.sumOf { it.attempts }
    val totalMisses: Int get() = sessions.sumOf { it.misses }
    val totalSwishes: Int get() = sessions.sumOf { it.swishes }
    val shootingPercentage: Int
        get() = if (totalAttempts == 0) 0 else ((totalMakes.toDouble() / totalAttempts) * 100).toInt()
    val sessionsCompleted: Int get() = sessions.size

    val favoriteDrill: String
        get() = sessions.groupBy { it.drill }.maxByOrNull { it.value.size }?.key ?: "No sessions yet"

    val todayMakes: Int
        get() = sessions.filter { isToday(it.endDateMillis) }.sumOf { it.makes }

    val todayAttempts: Int
        get() = sessions.filter { isToday(it.endDateMillis) }.sumOf { it.attempts }

    val weeklySessionsCompleted: Int
        get() = sessions.count { isThisWeek(it.endDateMillis) }

    val bestSession: WorkoutSession?
        get() = sessions.maxByOrNull { it.percentage }

    val mostRecentSession: WorkoutSession?
        get() = sessions.maxByOrNull { it.endDateMillis }

    val currentFocus: String
        get() = when {
            shootingPercentage < targetFgGoal -> "Improve shot consistency"
            weeklySessionsCompleted < weeklySessionGoal -> "Complete more sessions this week"
            else -> "Keep building momentum"
        }

    private fun readSessions(): List<WorkoutSession> = runCatching {
        val array = JSONArray(preferences.getString(KEY_SESSIONS, "[]"))
        buildList {
            for (index in 0 until array.length()) {
                val json = array.getJSONObject(index)
                add(
                    WorkoutSession(
                        id = json.getString("id"),
                        drill = json.getString("drill"),
                        makes = json.getInt("makes"),
                        misses = json.getInt("misses"),
                        swishes = json.getInt("swishes"),
                        attempts = json.getInt("attempts"),
                        startDateMillis = json.getLong("startDateMillis"),
                        endDateMillis = json.getLong("endDateMillis"),
                    ),
                )
            }
        }
    }.getOrDefault(emptyList())

    private fun persistSessions() {
        val array = JSONArray()
        sessions.forEach { session ->
            array.put(
                JSONObject()
                    .put("id", session.id)
                    .put("drill", session.drill)
                    .put("makes", session.makes)
                    .put("misses", session.misses)
                    .put("swishes", session.swishes)
                    .put("attempts", session.attempts)
                    .put("startDateMillis", session.startDateMillis)
                    .put("endDateMillis", session.endDateMillis),
            )
        }
        preferences.edit { putString(KEY_SESSIONS, array.toString()) }
    }

    private fun isToday(timeMillis: Long): Boolean {
        val now = Calendar.getInstance()
        val date = Calendar.getInstance().apply { timeInMillis = timeMillis }
        return now.get(Calendar.ERA) == date.get(Calendar.ERA) &&
            now.get(Calendar.YEAR) == date.get(Calendar.YEAR) &&
            now.get(Calendar.DAY_OF_YEAR) == date.get(Calendar.DAY_OF_YEAR)
    }

    private fun isThisWeek(timeMillis: Long): Boolean {
        val startOfWeek = Calendar.getInstance().apply {
            set(Calendar.DAY_OF_WEEK, firstDayOfWeek)
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        val endOfWeek = (startOfWeek.clone() as Calendar).apply {
            add(Calendar.WEEK_OF_YEAR, 1)
        }
        return timeMillis >= startOfWeek.timeInMillis && timeMillis < endOfWeek.timeInMillis
    }

    private companion object {
        const val KEY_SESSIONS = "squatch.sessions"
        const val KEY_DAILY_SHOT_GOAL = "squatch.dailyShotGoal"
        const val KEY_WEEKLY_SESSION_GOAL = "squatch.weeklySessionGoal"
        const val KEY_TARGET_FG_GOAL = "squatch.targetFGGoal"
    }
}
